vdf/src/hashchain.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+//! Instantiation of the hash chain delay function.
+
+use crate::{VDFError, VDF};
+use ark_serialize::{CanonicalDeserialize, CanonicalSerialize};
+use ark_std::vec::Vec;
+use core::marker::PhantomData;
+use sha3::Digest;
+
+/// Glorified bool type
+type VerificationResult = Result<(), ()>;
+
+#[derive(
+    Copy,
+    Clone,
+    Debug,
+    Default,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    CanonicalSerialize,
+    CanonicalDeserialize,
+)]
+/// Public parameter for MinRoot DF,
+pub struct HashChainParam {
+    /// Indicates the number of iterations
+    pub difficulty: u64,
+}
+
+#[derive(Copy, Debug, Clone)]
+/// Dummy struct for MinRoot delay function.
+pub struct HashChain;
+
+impl VDF for HashChain {
+    type PublicParameter = HashChainParam;
+    type Proof = [u8; 32];
+    type Input = [u8; 32];
+    type Output = [u8; 32];
+
+    fn setup<R: ark_std::rand::CryptoRng + ark_std::rand::RngCore>(
+        difficulty: u64,
+        prng: Option<&mut R>,
+    ) -> Result<Self::PublicParameter, VDFError> {
```

Comment:
Is it possible to make `setup` and `eval` infallible (not return a `Result`)? It looks like they can't fail in either implementation

not really a big deal: we can of course always handle the failure in HotShot

Reply:
I prefer to keep it. Error may occur when we are implementing Verifiable DF.